### ISSUE TITLE
Vim mode multi-key mappings

### ISSUE THREAD

**Issue by cjfuller** [zed-industries/zed#5337]
**Is your feature request related to a problem? Please describe.**
I normally remap keys in vim or vim-mode in other editors so that I can use `jk` in insert mode to exit insert mode instead of escape, but I don't see any way to do this from the docs, and it did not work to try to hack it with:
```
    {
        "context": "Editor && vim_mode == insert && vim_operator == j",
        "bindings": {
            "k": "vim::NormalBefore"
        }
    }
``` 

**Describe the solution you'd like**
I'd like to somehow be able to bind j then k to Escape in insert mode.

**Describe alternatives you've considered**
I tried the config hack above to no avail, and I also considered retraining to ctrl-c, which is the default alternate binding, but I'd rather be able to use something on the home row.

**Mockup**
For example, I'd love to be able to write the following:
```
    {
        "context": "Editor && vim_mode == insert",
        "bindings": {
            "jk": "vim::NormalBefore"
        }
    }
```

**Comment by LORE-D-NATO** [comment]
I have ```kj``` set as my escape binding and it makes exiting insert mode after typing a ```k``` sort of awkward with the 1000ms delay. Shouldn't ```kk``` type a ```k``` and then wait for another input?

**Comment by ConradIrwin** [comment]
@LORE-D-NATO If you just keep typing it should do the right thing, for example if you type `koala` without pausing after the `k` it should insert the k correctly. Please let me know if you're not observing that!

There does seem to be a small bug right now where if you type `kk` it doesn't type the first `k` until the timeout has passed either; I'll take a look at that.

**Comment by LORE-D-NATO** [comment]
@ConradIrwin Yes, typing a second letter other than `k` works as intended. The bug when typing `kk` is what I was referencing.

**Comment by jparr721** [comment]
I didn't want to resurrect this for another individual issue, so let me know if this is the inappropriate place to put this, but I am having trouble getting certain chained functionality similar to above, specifically `ciw` or "change inner word". I grabbed a keymap a year or so ago and have been using that, but the one critical piece keeping me in regular vim/vscode is the ability to use the `c` commands as I expect, and most specifically `ciw`. Given the following config, I am trying to understand how I might implement this? I have been googling this issue for months to avoid adding yet-another one-off request, but I am truly stumped as to how to do this with the present functionality. Is there any way I can get this to work as I expect? You can see my half-hearted attempt to add the feature in there, but it's definitely not right. If there's docs I'm happy to read those. I've been using the provided [bindings](https://docs.zed.dev/configuration/key-bindings), but I am not sure how I would replace the current word.

Something like this:
```
"c i w": "editor::DeleteToNextWordEnd",
```

Works great, but something like this:
```
      "c i w": [
        "editor::MoveToPreviousWordStart",
        "editor::DeleteToNextWordStart",
        "vim::SwitchMode",
        "Insert"
      ],
```
does not. Am I doing something horribly wrong here?

**Comment by ConradIrwin** [comment]
@jparr721 There should be (but I don't think there is) an issue for what you're talking about, which is being able to bind one keystroke to many actions (or more keystrokes). It is in the backlog of the Vim channel notes https://zed.dev/channel/vim-393, but would be good to have tracked on Github too.

That said, we do already support things like `ciw`. If you're looking at the binding file, you can see that  `c` pushes the "Change" operator: https://github.com/zed-industries/zed/blob/main/assets/keymaps/vim.json#L296

Then `i`  pushes the around operator: https://github.com/zed-industries/zed/blob/main/assets/keymaps/vim.json#L459, and finally `w` selects the word object: https://github.com/zed-industries/zed/blob/main/assets/keymaps/vim.json#L376.

Its a little convoluted, but that way we don't have to have separate bindings for each action and object.

In order for any of this work you do have to use Zed in Vim mode `"vim_mode": true` in your settings in addition to having the keymap; as that enables all the bookkeeping machinery.  (Also if you copied the keymap a year ago, I'd suggest deleting it and starting afresh!)

**Comment by jparr721** [comment]
The new keymaps worked like a charm! Thanks for the followup here.

**Comment by eneiford** [comment]
can you change the 1000ms delay time waiting on a second keypress (especially for a specific context)?  using jk is really awkward with the 1000ms delay when actually typing a "j" or "k".  If this delay setting was configurable it would make for a smoother experience.

**Comment by ConradIrwin** [comment]
@eneiford not right now, but if you’d like to pair with me on building it, feel free to book time at https://calendly.com/conradirwin/pairing. 

We should just be able to add a new setting for this.

**Comment by mre** [comment]
@ConradIrwin, I have a similar (but slightly different) use-case for reducing the 1000ms delay time.
(Did not want to open a new issue right away.)

I normally use `cmd-j` and `cmd-k` to switch between panes.  
  
```json
"bindings": {
    "cmd-j": "pane::ActivatePrevItem",
    "cmd-k": "pane::ActivateNextItem",
}
```

While `cmd-j` works as expected, `cmd-k` is slow because it waits for a second key press. I think if I set the delay time to 0, it would be instant. Alternatively, is there a way to disable multi-key mappings and/or set all the other `cmd-k X` mappings to `null`?

My current workaround is

```json
"bindings": {
    "cmd-[": "pane::ActivatePrevItem",
    "cmd-]": "pane::ActivateNextItem",
}
```

**Comment by ConradIrwin** [comment]
There's no currently easy way to do this unfortunately.

A few things we could consider improving on the Zed side that might help:
* Moving all the `cmd-k X` bindings to a top-level context so that user-bindings at a lower-level context would take priority
* Have a way to give user-bindings precedence over internal bindings
* Make the delay so much shorter that you don't care anymore

I think 3. should be easy, and is probably a good place to start; but 2. would be nice to have too. (1. seems further away from feasible)

**Comment by ConradIrwin** [comment]
Please do! 

Or if you'd like to just fix them PRs are also welcome – if you'd like to work together you can book time here: https://calendly.com/conradirwin/pairing

**Comment by cyfyifanchen** [comment]
```
  {
    "context": "Editor && vim_mode == normal && !menu",
    "bindings": {
      "j": ["g j"],
      "k": ["g", "k"]
    }
  }
```

Anyone knows how to make the above mapping work? I've tried various combination, don't seem to work at all.

**Comment by ConradIrwin** [comment]
@cyfyifanchen You either need to use the workspace::SendKeystrokes action to convert one set of keys to another: https://zed.dev/docs/key-bindings#remapping-keys

Or (for this specific case), you can use the target actions directly (https://github.com/zed-industries/zed/blob/2f6cb49d841284e2068497f1564320da108be45f/assets/keymaps/vim.json#L195):

```
      "j": [
        "vim::Down",
        {
          "displayLines": true
        }
      ],
      "k": [
        "vim::Up",
        {
          "displayLines": true
        }
      ],
```

**Comment by cyfyifanchen** [comment]
@ConradIrwin beautiful!

**Comment by ConradIrwin** [comment]
@venkatb-zelar yes!

You could do something like this: 
```
{
  "context": "vim_mode == insert",
  "bindings": {
    "j k": ["workspace::SendKeystrokes", "escape command-s"]
  }
}
```

You could also look into the Zed `"autosave": "on_focus_change"` setting, which I use to avoid having to think about saving too hard.

**Comment by nmsobri** [comment]
> I've added support for `jk` in insert-mode in the latest nightly:
> 
> * You'll need to set up bindings like:
> 
> ```json
> [
>   {
>     "context": "Editor && vim_mode == insert",
>     "bindings": {
>       "j k": "vim::NormalBefore"
>     }
>   }
> ]
> ```
> 
> * Once you have that, typing `jk` in insert mode will act like `<esc>`
> * If you type `j` and then wait 1000ms, it will type `j`
> * If you type `j` and then any other key, it will type the `j` just before the next key.
> 
> This works similarly in normal mode, for example if you have:
> 
> ```json
> [
>   {
>     "bindings": {
>       "shift-escape": null,
>       "ctrl-enter": "menu::ShowContextMenu",
>       ", w": "workspace::Save"
>     }
>   }
> ]
> ```
> 
> * If you type `,w` then it will save.
> * If you type `,` and then wait 1000ms it will do the default (repeat the most recent `tf` command in reverse).
> * If you type `,` and then another key, it will do the default just before the next key
> 
> With that done, I think we can close this out.
> 
> There is a related feature request to support mapping from one sequence of keys to another sequence of keys, but we should track that separately!

still not a viable solution.. I mean, if you really want to type `jk`, you need to type `j` and wait for 1 sec to type `k`, or else Zed will interpret as escape to normal mode.

is there a way to reduce vim delay timeout? normally i set this value in .vimrc to 400ms

**Comment by tednaaa** [comment]
any updates? in neovim I use 250 timeout, and default 1000 is too much for me

**Comment by ConradIrwin** [comment]
If you want to work on this with me: https://calendly.com/conradirwin/pairing

**Comment by lungarella-raffaele** [comment]
Any updates on configuring the timeout?
